File: src/agent/tool_search.py
"""渐进披露(工具搜索)纯逻辑 + 会话级激活状态。

对齐 dashboard 已实现的模式(见 dashboard/electron/main/kernel/tool-search.ts):
连接器/远程工具装多后不再每轮把全部远程工具定义塞给模型, 而是先发内置工具 +
`search_tools`, 由模型按需检索, 命中后把远程工具「激活」进本对话(会话内粘住、
重启保留)。全部判定/检索/文案为纯函数, 便于离线单测; 激活持久化优先走
`session_meta`(见 storage.get_active_tools/set_active_tools), 不可用时进程内内存兜底。
"""

import logging
import os
import threading
from collections.abc import Iterable, Mapping
from dataclasses import dataclass
from enum import Enum
from typing import Any

logger = logging.getLogger("agent.tool_search")

# search_tools 注册名(内置工具通道, 与 ask_user 同级)
SEARCH_TOOLS_NAME = "search_tools"
# auto 模式默认阈值: 远程工具数 > 40 时启用渐进披露
DEFAULT_TOOL_SEARCH_THRESHOLD = 40
# search_tools 的 limit 缺省值与夹紧范围(与 dashboard 一致)
SEARCH_TOOLS_DEFAULT_LIMIT = 8
SEARCH_TOOLS_MAX_LIMIT = 20

ENV_TOOL_SEARCH = "AGENT_TOOL_SEARCH"
ENV_TOOL_SEARCH_THRESHOLD = "AGENT_TOOL_SEARCH_THRESHOLD"

# 远程工具来源(可渐进): MCP server / 插件 / 市场远程; 核心来源恒注入
REMOTE_SOURCES = frozenset({"mcp", "plugin", "market"})
CORE_SOURCES = frozenset({"builtin", "skill"})


class ToolSearchMode(str, Enum):
    """`AGENT_TOOL_SEARCH` 取值: auto(默认)/always/off。"""

    AUTO = "auto"
    ALWAYS = "always"
    OFF = "off"


def parse_tool_search_mode(value: Any) -> ToolSearchMode:
    """解析模式字符串; 非法值告警并回退 auto(缺省即默认)。"""
    raw = getattr(value, "value", value)  # 兼容直接传入 ToolSearchMode 成员(str Enum)
    text = str(raw or "").strip().lower()
    if text in (ToolSearchMode.AUTO.value, ToolSearchMode.ALWAYS.value, ToolSearchMode.OFF.value):
        return ToolSearchMode(text)
    if text:
        logger.warning(
            f"{ENV_TOOL_SEARCH} 非法值 {raw!r}, 回退 auto(工具多时渐进)")
    return ToolSearchMode.AUTO


def parse_tool_search_threshold(value: Any, default: int = DEFAULT_TOOL_SEARCH_THRESHOLD) -> int:
    """解析阈值; 缺省/非法回退 default(40), 负值视为非法。"""
    text = str(value).strip() if value is not None else ""
    if not text:
        return default
    try:
        number = int(text)
    except (TypeError, ValueError):
        logger.warning(f"{ENV_TOOL_SEARCH_THRESHOLD} 非法值 {value!r}, 回退默认 {default}")
        return default
    if number < 0:
        logger.warning(f"{ENV_TOOL_SEARCH_THRESHOLD} 非法值 {value!r}, 回退默认 {default}")
        return default
    return number


def tool_search_mode_from_env(env: Mapping[str, str] | None = None) -> ToolSearchMode:
    """从环境变量读取模式(缺省 auto)。"""
    source = os.environ if env is None else env
    return parse_tool_search_mode(source.get(ENV_TOOL_SEARCH))


def tool_search_threshold_from_env(env: Mapping[str, str] | None = None) -> int:
    """从环境变量读取阈值(缺省 40)。"""
    source = os.environ if env is None else env
    return parse_tool_search_threshold(source.get(ENV_TOOL_SEARCH_THRESHOLD))


def should_use_progressive(mode: ToolSearchMode | str, remote_count: int,
                           threshold: int = DEFAULT_TOOL_SEARCH_THRESHOLD) -> bool:
    """是否启用渐进披露: off 恒否; always 恒是; auto 仅当远程工具数 > 阈值(边界 40/41)。"""
    normalized = mode if isinstance(mode, ToolSearchMode) else parse_tool_search_mode(mode)
    if normalized is ToolSearchMode.OFF:
        return False
    if normalized is ToolSearchMode.ALWAYS:
        return True
    return int(remote_count) > int(threshold)


def is_remote_tool(name: str, source_hint: str | None = None) -> bool:
    """远程工具判定(可渐进): 来源优先(builtin/skill→核心, mcp/plugin/market→远程)。

    来源缺失/未知时按暴露名兜底: 平台轨 `platform__*`、MCP 重名前缀 `mcp__*`、
    市场远程 `market:*` 一律视为远程; 其余按核心处理。
    """
    hint = str(source_hint or "").strip().lower()
    if hint:
        return hint in REMOTE_SOURCES
    text = str(name or "")
    return text.startswith("platform__") or text.startswith("mcp__") or text.startswith("market:")


@dataclass(frozen=True)
class ToolSearchEntry:
    """检索条目: 远程工具名 + 描述 + 来源(连接器/插件名)。"""

    name: str
    description: str = ""
    source: str = ""


@dataclass(frozen=True)
class ToolSearchHit:
    """检索命中: score 用于排序与测试断言(名称精确 100 / 名称包含 10 / 来源 5 / 描述 2)。"""

    name: str
    description: str
    source: str
    score: int


def clamp_limit(limit: Any, default: int = SEARCH_TOOLS_DEFAULT_LIMIT,
                max_limit: int = SEARCH_TOOLS_MAX_LIMIT) -> int:
    """limit 夹紧: 非数值/缺省取 default, 结果恒在 1..max_limit。"""
    try:
        number = int(limit)
    except (TypeError, ValueError):
        number = default
    return max(1, min(number, max_limit))


def search_remote_tools(entries: Iterable[ToolSearchEntry], query: str,
                        limit: Any = SEARCH_TOOLS_DEFAULT_LIMIT,
                        max: int = SEARCH_TOOLS_MAX_LIMIT) -> tuple[list[ToolSearchHit], str]:
    """远程工具检索(纯函数)。

    query 按空白分词, 每词在「名称 / 来源 / 描述」中做不区分大小写的子串匹配并计分
    (名称精确 +100、名称包含 +10、来源包含 +5、描述包含 +2), 至少命中一词才入选;
    按分数降序、同分按名称升序稳定排序; 返回 (命中列表, 文案), 未命中/空查询给
    可行动提示。
    """
    terms = [t for t in str(query or "").lower().split() if t]
    candidates = list(entries or [])
    if not terms:
        return [], "请提供搜索关键词（工具名/功能描述/连接器名，空格分隔多个词）。"
    hits: list[ToolSearchHit] = []
    for entry in candidates:
        name = str(getattr(entry, "name", "") or "")
        if not name:
            continue
        lowered_name = name.lower()
        source = str(getattr(entry, "source", "") or "").lower()
        description = str(getattr(entry, "description", "") or "").lower()
        score = 0
        for term in terms:
            if lowered_name == term:
                score += 100
            elif term in lowered_name:
                score += 10
            if source and term in source:
                score += 5
            if term in description:
                score += 2
        if score > 0:
            hits.append(ToolSearchHit(
                name=name,
                description=str(getattr(entry, "description", "") or ""),
                source=str(getattr(entry, "source", "") or ""),
                score=score,
            ))
    hits.sort(key=lambda hit: (-hit.score, hit.name))
    hits = hits[:clamp_limit(limit, max_limit=max)]
    if not hits:
        return [], (f"未找到匹配「{query}」的远程工具（当前可用 {len(candidates)} 个）。"
                    "可换关键词重试，或直接用现有工具完成任务。")
    return hits, format_search_result(hits)


def format_search_result(hits: Iterable[ToolSearchHit]) -> str:
    """search_tools 命中输出: 逐条「名称 —— [连接器 x] 描述」, 并说明已激活。"""
    items = list(hits or [])
    lines = [f"找到 {len(items)} 个匹配的远程工具（已激活，下一轮对话可直接调用）："]
    for hit in items:
        tag = f"[连接器 {hit.source}] " if hit.source else ""
        desc = " ".join(str(hit.description or "").split()) or "(无描述)"
        lines.append(f"{hit.name} —— {tag}{desc}")
    return "\n".join(lines)


def progressive_hint(active_names: Iterable[str]) -> str:
    """渐进模式下追加到系统提示末尾的说明(文案与 dashboard 一致)。"""
    names = [str(name).strip() for name in (active_names or []) if str(name).strip()]
    listing = "、".join(names) if names else "无"
    return (
        "连接器的更多工具需先用 `search_tools` 搜索；已激活："
        + listing
        + "。需要远程能力（远程终端/设备/市场等）而当前工具列表没有时，先搜索再调用。"
    )


class ToolActivationStore:
    """会话激活集: session_meta 持久化优先, 不可用时进程内内存兜底。

    - 键为「对话根」(`RunContext.conversation_id`, web/钉钉/子代理一致);
    - load: 进程内缓存命中直接返回; 未命中读 `storage.get_active_tools`(JSON 数组),
      storage 缺失/读取异常 → 返回空集且不缓存(下次可重试);
    - activate: 合并去重后先更新进程内缓存, 再写穿 storage(失败仅告警, 内存已生效);
    - 仅 `search_tools` 增加; 远程工具暂不可用时注入前过滤, 不清理持久化
      (server 恢复后自动重新可用)。
    """

    _lock = threading.Lock()
    _cache: dict[str, frozenset[str]] = {}

    @staticmethod
    def _normalize(conversation_id: Any) -> str:
        return str(conversation_id or "").strip()

    @classmethod
    def load(cls, conversation_id: Any, storage: Any = None) -> frozenset[str]:
        """读取该对话已激活的远程工具名集合。"""
        cid = cls._normalize(conversation_id)
        if not cid:
            return frozenset()
        with cls._lock:
            cached = cls._cache.get(cid)
        if cached is not None:
            return cached
        if storage is None or not hasattr(storage, "get_active_tools"):
            return frozenset()
        try:
            raw = storage.get_active_tools(cid)
        except Exception as e:  # noqa: BLE001 — 存储异常回退内存, 不阻断对话
            logger.warning(f"读取会话激活工具失败(内存兜底): {e}")
            return frozenset()
        names = frozenset(str(n) for n in (raw or []) if str(n).strip())
        with cls._lock:
            cls._cache[cid] = names
        return names

    @classmethod
    def activate(cls, conversation_id: Any, names: Iterable[str], storage: Any = None) -> frozenset[str]:
        """把命中的工具名并入激活集并持久化, 返回最新激活集。"""
        cid = cls._normalize(conversation_id)
        added = {str(name).strip() for name in (names or []) if str(name).strip()}
        if not cid:
            return frozenset()
        if not added:
            return cls.load(cid, storage)
        merged = set(cls.load(cid, storage)) | added
        value = frozenset(merged)
        with cls._lock:
            cls._cache[cid] = value
        if storage is not None and hasattr(storage, "set_active_tools"):
            try:
                storage.set_active_tools(cid, sorted(merged))
            except Exception as e:  # noqa: BLE001 — 持久化失败内存仍生效
                logger.warning(f"持久化会话激活工具失败(内存已生效): {e}")
        return value

    @classmethod
    def reset_cache(cls, conversation_id: Any = None) -> None:
        """清空进程内缓存(测试/运维用); 不传 conversation_id 清全部。"""
        with cls._lock:
            if conversation_id is None:
                cls._cache.clear()
            else:
                cls._cache.pop(cls._normalize(conversation_id), None)

File: desktop/electron/main/kernel/tool-search.ts
import type { ToolDefinition } from './types'

/**
 * 本地 agent 渐进披露（工具搜索）纯逻辑：
 * 连接器装多后不再每轮把全部远程工具定义发给模型，而是先发内置工具 + search_tools，
 * 由模型按需调用 search_tools 检索，命中后把远程工具「激活」进本会话（会话内粘住）。
 * 全部为纯函数/工厂，无 IO，便于离线单测；配置读取与会话持久化由 ipc 层注入。
 */

export type ToolSearchMode = 'auto' | 'always' | 'off'

/** auto 模式的默认阈值：已注册远程工具数 > 40 时启用渐进披露 */
export const TOOL_SEARCH_THRESHOLD = 40
/** search_tools 的 limit 缺省值与夹紧范围 */
export const SEARCH_TOOLS_DEFAULT_LIMIT = 8
export const SEARCH_TOOLS_MAX_LIMIT = 20

/** search_tools 注册名（渐进模式下由内置工具通道恒发） */
export const SEARCH_TOOLS_NAME = 'search_tools'

/**
 * 远程工具判定：连接器注册进 registry 的工具——
 * MCP 的 `mcp__<server>__<tool>`（mcp.ts mcpToolName）与市场远程 `market:<name>`（market-tools.ts）。
 * 禁用/卸载的连接器不会被注册，因此天然不计入。
 */
export function isRemoteTool(name: string): boolean {
  return name.startsWith('mcp__') || name.startsWith('market:')
}

/** 配置值规范化：存储里出现脏值时回退 auto，调用方无需处理非法入参 */
export function normalizeToolSearchMode(value: unknown): ToolSearchMode {
  return value === 'always' || value === 'off' ? value : 'auto'
}

/** 是否启用渐进披露：off 恒否；always 恒是；auto 仅当已注册远程工具数 > 阈值 */
export function shouldUseProgressive(
  mode: ToolSearchMode,
  remoteCount: number,
  threshold: number = TOOL_SEARCH_THRESHOLD
): boolean {
  if (mode === 'off') return false
  if (mode === 'always') return true
  return remoteCount > threshold
}

/** 检索条目：registry 里的工具名与描述（仅需这两项） */
export interface ToolSearchEntry {
  name: string
  description: string
}

/** 从工具名解析所属连接器：mcp__<server>__<tool> → server，market:<name> → market，无法判定回空串 */
export function connectorOf(name: string): string {
  if (name.startsWith('mcp__')) {
    const rest = name.slice('mcp__'.length)
    const at = rest.indexOf('__')
    return at > 0 ? rest.slice(0, at) : ''
  }
  if (name.startsWith('market:')) return 'market'
  return ''
}

/** 检索命中条目：score 用于排序与测试断言 */
export interface ToolSearchHit {
  name: string
  description: string
  score: number
}

/** limit 夹紧：非有限值/缺省取 8，结果恒在 1..20 */
export function clampLimit(limit?: number): number {
  const n = typeof limit === 'number' && Number.isFinite(limit)
    ? Math.floor(limit)
    : SEARCH_TOOLS_DEFAULT_LIMIT
  return Math.min(Math.max(n, 1), SEARCH_TOOLS_MAX_LIMIT)
}

/**
 * 远程工具检索（纯函数）：
 * query 按空白分词，每个词在「名称 / 连接器 / 描述」中做不区分大小写的子串匹配并计分
 * （名称命中权重最高、连接器其次、描述最低），至少命中一个词才入选；
 * 按分数降序、同分按名称升序稳定排序，截取 limit（夹紧 1..20，缺省 8）。
 * 空查询/空白查询返回空数组（由调用方提示补充关键词）。
 */
export function searchRemoteTools(
  entries: readonly ToolSearchEntry[],
  query: string,
  limit?: number
): ToolSearchHit[] {
  const terms = String(query ?? '')
    .toLowerCase()
    .split(/\s+/)
    .filter(Boolean)
  if (terms.length === 0) return []
  const hits: ToolSearchHit[] = []
  for (const entry of entries) {
    if (!entry || typeof entry.name !== 'string' || !entry.name) continue
    const name = entry.name.toLowerCase()
    const connector = connectorOf(entry.name).toLowerCase()
    const description = String(entry.description ?? '').toLowerCase()
    let score = 0
    for (const term of terms) {
      if (name === term) score += 100
      else if (name.includes(term)) score += 10
      if (connector && connector.includes(term)) score += 5
      if (description.includes(term)) score += 2
    }
    if (score > 0) hits.push({ name: entry.name, description: entry.description, score })
  }
  hits.sort((a, b) => b.score - a.score || (a.name < b.name ? -1 : a.name > b.name ? 1 : 0))
  return hits.slice(0, clampLimit(limit))
}

/** search_tools 输出：逐条「名称 —— [连接器 x] 描述」，并说明命中工具已激活、下一轮可直接调用 */
export function formatSearchResult(hits: readonly ToolSearchHit[]): string {
  const lines = hits.map((hit) => {
    const connector = connectorOf(hit.name)
    const tag = connector ? `[连接器 ${connector}] ` : ''
    const desc = String(hit.description ?? '').replace(/\s+/g, ' ').trim()
    return `${hit.name} —— ${tag}${desc || '(无描述)'}`
  })
  return [`找到 ${hits.length} 个匹配的远程工具（已激活，下一轮对话可直接调用）：`, ...lines].join('\n')
}

/** 渐进模式下追加到系统提示末尾的说明：告知先搜索、并列出本会话已激活的工具 */
export function progressiveHint(activeNames: readonly string[]): string {
  const names = activeNames.map((n) => String(n ?? '').trim()).filter(Boolean)
  const list = names.length > 0 ? names.join('、') : '无'
  return (
    '连接器的更多工具需先用 `search_tools` 搜索；已激活：' +
    list +
    '。需要远程能力（远程终端/设备/市场等）而当前工具列表没有时，先搜索再调用。'
  )
}

/** 每轮发送给模型的工具组装：非渐进=全量；渐进=非远程工具（含内置与 search_tools）+ 本会话已激活且仍注册的远程工具 */
export function selectRoundTools<T extends { function: { name: string } }>(
  all: readonly T[],
  progressive: boolean,
  activeNames: readonly string[]
): T[] {
  if (!progressive) return [...all]
  const active = new Set(activeNames)
  return all.filter((t) => !isRemoteTool(t.function.name) || active.has(t.function.name))
}

export interface SearchToolsDeps {
  /** 当前已注册的远程工具（禁用/卸载的连接器不在其中） */
  listRemote(): ToolSearchEntry[]
  /** 把命中的工具名合并进本会话激活集并持久化，返回最新激活集 */
  activate(names: string[], sessionId: string): string[]
}

/** 构造 search_tools 工具（kind: read，不触发权限确认）：检索 → 激活 → 文本清单 */
export function createSearchToolsTool(deps: SearchToolsDeps): ToolDefinition {
  return {
    name: SEARCH_TOOLS_NAME,
    description:
      '在已启用连接器的远程工具（MCP / 市场远程能力）中按关键词搜索。命中的工具会立即激活，下一轮对话即可直接调用；' +
      '当需要远程能力（远程终端/设备/市场等）而当前工具列表里没有时，先用本工具搜索。',
    kind: 'read',
    parameters: {
      type: 'object',
      properties: {
        query: { type: 'string', description: '搜索关键词（工具名/功能描述/连接器名，空格分隔多个词）' },
        limit: {
          type: 'integer',
          description: `最多返回条数（1-${SEARCH_TOOLS_MAX_LIMIT}），缺省 ${SEARCH_TOOLS_DEFAULT_LIMIT}`
        }
      },
      required: ['query']
    },
    async execute(args, ctx) {
      const query = typeof args.query === 'string' ? args.query.trim() : ''
      if (!query) return { ok: false, output: '参数 query 必须为非空字符串' }
      try {
        const entries = deps.listRemote()
        const hits = searchRemoteTools(
          entries,
          query,
          typeof args.limit === 'number' ? args.limit : undefined
        )
        if (hits.length === 0) {
          return {
            ok: true,
            output: `未找到匹配「${query}」的远程工具（当前已注册 ${entries.length} 个）。可换关键词重试，或直接用现有工具完成任务。`
          }
        }
        deps.activate(hits.map((h) => h.name), ctx.sessionId)
        return { ok: true, output: formatSearchResult(hits) }
      } catch (err) {
        return { ok: false, output: `工具搜索失败: ${err instanceof Error ? err.message : String(err)}` }
      }
    }
  }
}
